src/MinterGateway.sol

```diff
@@ -602,10 +602,70 @@ contract MinterGateway is IMinterGateway, ContinuousIndexing, ERC712Extended {
+
+        uint32 penaltyRate_ = penaltyRate();
+
+        if (penaltyRate_ == 0) return (0, 0);
+
+        MinterState storage minterState_ = _minterStates[minter_];
+
+        (uint40 missedIntervals_, uint40 missedUntil_) = _getMissedCollateralUpdateParameters(
+            minterState_.updateTimestamp,
+            minterState_.penalizedUntilTimestamp,
+            updateCollateralInterval()
+        );
+
+        unchecked {
+            // As an edge case precaution, cap the penalty principal to type(uint112).max.
+            uint112 missedIntervalPenaltyPrincipal_ = UIntMath.bound112(
+                (uint256(principalOfActiveOwedM_) * missedIntervals_ * penaltyRate_) / ONE
+            );
+
+            principalOfActiveOwedM_ = UIntMath.bound112(
+                uint256(principalOfActiveOwedM_) + uint256(missedIntervalPenaltyPrincipal_)
+            );
+
+            missedIntervalPenalty_ = _getPresentAmount(missedIntervalPenaltyPrincipal_);
+        }
+
+        uint256 maxAllowedActiveOwedM_ = maxAllowedActiveOwedMOf(minter_);
+
+        // If the minter's max allowed active owed M is greater than `type(uint240).max`, then it's definitely greater
+        // than the max possible active owed M for the minter, which is capped at `type(uint240).max`.
+        if (maxAllowedActiveOwedM_ >= type(uint240).max) return (missedIntervalPenalty_, 0);
+
+        // NOTE: Round the principal down in favor of the protocol since this is a max applied to the minter.
+        uint112 principalOfMaxAllowedActiveOwedM_ = _getPrincipalAmountRoundedDown(uint240(maxAllowedActiveOwedM_));
+
+        // If the minter is not undercollateralized, then no penalty is imposed.
+        if (principalOfMaxAllowedActiveOwedM_ >= principalOfActiveOwedM_) return (missedIntervalPenalty_, 0);
 
-        return (penaltyPrincipal_ == 0) ? 0 : _getPresentAmount(penaltyPrincipal_);
+        unchecked {
+            uint112 principalOfExcessOwedM_ = principalOfActiveOwedM_ - principalOfMaxAllowedActiveOwedM_;
+
+            // NOTE: `block.timestamp >= missedUntil_` since `_getMissedCollateralUpdateParameters` ensures that
+            //       `missedUntil_` is only up to `block.timestamp`.
+            //
+            // NOTE: `block.timestamp - missedUntil_` will never be larger than `updateCollateralInterval_` since this
+            //       is only computed after `_getMissedCollateralUpdateParameters`, which ensures that the
+            //       `missedUntil_` is within one `updateCollateralInterval_` of the `block.timestamp`.
+            //
+            // NOTE: `updateCollateralInterval()` never equals 0, so the division is safe.
+            //       Its minimum is capped at `MIN_UPDATE_COLLATERAL_INTERVAL`.
+            uint112 undercollateralizedPenaltyPrincipal_ = UIntMath.bound112(
+                (penaltyRate_ *
+                    ((principalOfExcessOwedM_ * (uint40(block.timestamp) - missedUntil_)) /
```

Comment:
`block.timestamp` is not accurate, we use min validator timestamp for the selected most optimal solution, we can ask the user to pass min validator timestamp